### PR TITLE
Add expired detail into the introspect response.

### PR DESCRIPTION
**Pull request by Pushpalanka**
### Proposed changes in this pull request

Add the isExpired detail into the introspect response as below.

`{"user":{"username":"4453@wso2.com","tenant-domain":"carbon.super","realm":"PRIMARY"},"recoveryScenario":"LITE_SIGN_UP","recoveryStep":"CONFIRM_LITE_SIGN_UP","isExpired":true}`

### CODE REVIEW

File: components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/model/UserRecoveryData.java [hunk 1]
```diff
@@ -45,6 +47,13 @@ public UserRecoveryData(User user, String secret, Enum recoveryScenario) {
         this.recoveryScenario = recoveryScenario;
     }
 
+    public UserRecoveryData(User user, String secret, Enum recoveryScenario, Enum recoveryStep, boolean isExpired) {
```
Comment (somindatommy) [inline components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/model/UserRecoveryData.java:50]: Shall we add a new line after the method signature?

File: components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/store/UserRecoveryDataStore.java [hunk 2]
```diff
@@ -37,6 +38,21 @@ UserRecoveryData load(String code) throws
     UserRecoveryData load(User user) throws
             IdentityRecoveryException;
 
+    /**
+     * Load UserRecoveryData from the
```
Comment (somindatommy) [inline components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/store/UserRecoveryDataStore.java:42]: I think we might have missed some of the details here

File: components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/store/UserRecoveryDataStore.java [hunk 3]
```diff
@@ -37,6 +38,21 @@ UserRecoveryData load(String code) throws
...
+     *
+     * @param code                 Confirmation code
```
Comment (somindatommy) [inline components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/store/UserRecoveryDataStore.java:44]: Shall we end the description with a fullstop?

File: components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/signup/UserSelfRegistrationManager.java [hunk 4]
```diff
@@ -643,19 +643,44 @@ public UserRecoveryData introspectUserSelfRegistration(String code, String verif
...
 
-        return validateSelfRegistrationCode(code,verifiedChannelType,verifiedChannelClaim,properties);
+        return introspectUserSelfRegistration(false, code,verifiedChannelType,verifiedChannelClaim,properties);
+    }
+
+    /**
+     * Introspect the user self registration by validating the confirmation code, sets externally verified claims and
+     * return the details. Does not invalidate the code.
+     *
+     * @param skipExpiredCodeValidation   Skip confirmation code validation against expiration.
```
Comment (somindatommy) [inline components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/signup/UserSelfRegistrationManager.java:653]: Seems like there is a formatting issue. Shall we align the param descriptions and end all the descriptions with a full stop? Also add the return type info as well?

File: components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/store/JDBCRecoveryDataStore.java [hunk 5]
```diff
@@ -134,6 +135,12 @@ public UserRecoveryData load(User user, Enum recoveryScenario, Enum recoveryStep
     @Override
     public UserRecoveryData load(String code) throws IdentityRecoveryException {
 
+        return load(code, false);
+    }
+
+    @Override
+    public UserRecoveryData load(String code, boolean skipExpiryValidation) throws IdentityRecoveryException, NotImplementedException {
```
Comment (somindatommy) [inline components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/store/JDBCRecoveryDataStore.java:142]: How are we handling the skipping expiry validation?

File: components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/store/UserRecoveryDataStore.java (hunk 3 above)
Comment (Pushpalanka) [reply components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/store/UserRecoveryDataStore.java]: +1. Addressed.

File: components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/store/UserRecoveryDataStore.java (hunk 2 above)
Comment (Pushpalanka) [reply components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/store/UserRecoveryDataStore.java:42]: yes, added. Thanks for pointing out.

File: components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/model/UserRecoveryData.java (hunk 1 above)
Comment (Pushpalanka) [reply components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/model/UserRecoveryData.java:52]: Addressed following the generic coding style as mentioned.

File: components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/store/JDBCRecoveryDataStore.java (hunk 5 above)
Comment (Pushpalanka) [reply components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/store/JDBCRecoveryDataStore.java:142]: Code introspection method always calls setting this to 'true'. The rest of the existed use cases always set this to 'false' keeping the existing behavior, which does the expiry validation. There is little out of place login within the Data layer to check the expiry of the codes before return. This parameter is used there to make the switch.

**Review by senthalan — APPROVED** [review]

**Pull request merged**